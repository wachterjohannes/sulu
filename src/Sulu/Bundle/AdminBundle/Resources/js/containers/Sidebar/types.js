// @flow
import type {ComponentType} from 'react';

export type SidebarViewOptions = {};

export type SidebarView = ComponentType<SidebarViewOptions>;

export type Sizes = 'small' | 'middle' | 'large';

export type SidebarConfig = {
    view?: string,
    props?: Object,
    size?: Sizes,
};

const DEFAULT_SIZE = 'middle';

export {DEFAULT_SIZE};
