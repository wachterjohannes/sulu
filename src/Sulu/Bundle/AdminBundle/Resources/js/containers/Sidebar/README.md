The `Sidebar` is a configurable component which serves as a wrapper for multiple other component types. The
configuration can be set by using the `withSidebar` function which takes a React component as the first and a 
callback which returns the configuration object as the second parameter. In most cases the given component would be some
kind of a page or view component in which communicates with the `Sidebar`.

The configuration object describes how the Sidebar should behave and which view should be rendered inside.

Here is a basic usage example how `withSidebar` could be used:

```javascript static
const PageWithSidebar = withSidebar(Page, function() {
    return {
        view: 'test',
        props: {
            title: 'world',
        },
    };
});
```

In the above example a new sidebar-view was registered, used beside a component and passed some `props`.

Additionally you can pass which Sidebar `sizes` are supported by the component (one or more of `small`, `medium` or 
`large`) and a `defaultSize` when the sidebar will be initialized.
