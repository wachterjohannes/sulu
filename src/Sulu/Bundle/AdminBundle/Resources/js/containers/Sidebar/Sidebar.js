// @flow
import React from 'react';
import {observer} from 'mobx-react';
import classNames from 'classnames';
import sidebarStore from './stores/SidebarStore';
import sidebarViewRegistry from './registries/SidebarViewRegistry';
import sidebarStyles from './sidebar.scss';

type Props = {
    className?: string,
};

@observer
export default class Sidebar extends React.Component<Props> {
    render() {
        if (!sidebarStore.view) {
            return null;
        }

        const Component = sidebarViewRegistry.get(sidebarStore.view);
        const {
            className,
        } = this.props;

        const sidebarClass = classNames(
            sidebarStyles.sidebar,
            className
        );

        return (
            <aside className={sidebarClass}>
                <Component {...sidebarStore.props} />
            </aside>
        );
    }
}
