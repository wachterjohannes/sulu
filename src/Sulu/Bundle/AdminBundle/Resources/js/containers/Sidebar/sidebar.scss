@import '../Application/colors.scss';

$sidebarBorderColor: $silver;

.sidebar {
    border-left: 1px $sidebarBorderColor solid;
}
