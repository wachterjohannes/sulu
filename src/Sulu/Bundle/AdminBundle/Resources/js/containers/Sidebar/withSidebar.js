// @flow
import {autorun} from 'mobx';
import type {Component} from 'react';
import {buildHocDisplayName} from '../../services/react';
import type {SidebarConfig} from './types';
import sidebarStore from './stores/SidebarStore';

export default function withSidebar(
    Component: Class<Component<*, *>>,
    sidebar: () => SidebarConfig
) {
    const WithSidebarComponent = class extends Component {
        static hasSidebar = true;

        sidebarDisposer: Function;

        componentWillMount() {
            if (super.hasOwnProperty('sidebarDisposer')) {
                throw new Error('Component passed to withSidebar cannot declare a property called "sidebarDisposer".');
            }

            if (super.componentWillMount) {
                super.componentWillMount();
            }

            this.sidebarDisposer = autorun(() => {
                sidebarStore.setConfig(sidebar.call(this));
            });
        }

        componentWillUnmount() {
            if (super.componentWillUnmount) {
                super.componentWillUnmount();
            }

            this.sidebarDisposer();
        }
    };

    WithSidebarComponent.displayName = buildHocDisplayName('withSidebar', Component);

    return WithSidebarComponent;
}
