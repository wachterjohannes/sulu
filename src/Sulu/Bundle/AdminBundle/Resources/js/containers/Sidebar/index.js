// @flow
import Sidebar from './Sidebar';
import sidebarStore from './stores/SidebarStore';
import sidebarViewRegistry from './registries/SidebarViewRegistry';
import withSidebar from './withSidebar';

export default Sidebar;

export {
    sidebarStore,
    sidebarViewRegistry,
    withSidebar,
};
