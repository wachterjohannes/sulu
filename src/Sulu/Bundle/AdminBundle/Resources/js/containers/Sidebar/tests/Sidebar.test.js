/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import {render} from 'enzyme';
import Sidebar from '../Sidebar';
import sidebarStore from '../stores/SidebarStore';
import sidebarViewRegistry from '../registries/SidebarViewRegistry';

const component = (props) => (<h1>{props.title}</h1>);

jest.mock('../stores/SidebarStore', () => ({}));

jest.mock('../registries/SidebarViewRegistry', () => ({
    get: jest.fn(),
}));

test('Render correct sidebar view', () => {
    sidebarStore.view = 'preview';
    sidebarViewRegistry.get.mockReturnValue(component);

    const view = render(<Sidebar />);
    expect(view).toMatchSnapshot();
});

test('Render correct sidebar view with props', () => {
    sidebarStore.view = 'preview';
    sidebarStore.props = {title: 'Hello world'};
    sidebarViewRegistry.get.mockReturnValue(component);

    const view = render(<Sidebar />);
    expect(view).toMatchSnapshot();
});
