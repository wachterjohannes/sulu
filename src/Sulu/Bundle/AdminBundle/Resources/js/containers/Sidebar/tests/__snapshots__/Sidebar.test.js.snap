// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correct sidebar view 1`] = `
<aside
  class="sidebar"
>
  <h1 />
</aside>
`;

exports[`Render correct sidebar view with props 1`] = `
<aside
  class="sidebar"
>
  <h1>
    Hello world
  </h1>
</aside>
`;
